base/reflection.jl

```diff
@@ -600,6 +589,29 @@ Return `true` if `x` is an instance of an [`isbitstype`](@ref) type.
 """
 isbits(@nospecialize x) = isbitstype(typeof(x))
 
+"""
+    objectid(x) -> UInt
+
+Get a hash value for `x` based on object identity.
+
+If `x === y` then `objectid(x) == objectid(y)`, and usually when `x !== y`, `objectid(x) != objectid(y)`.
+
+See also [`hash`](@ref), [`IdDict`](@ref).
+"""
+function objectid(x)
+    # objectid is foldable iff it isn't a pointer.
+    # Sring, Symbol, Module, and DataType are safe because jl_objectid for them works by value.
+    if !ismutable(x) || x isa String || x isa Symbol || x isa Module
```

Comment:
`ismutable` doesn't feel right.

Consider

```
struct F
    v::Vector
end
```

That's not mutable but I don't think objectid is foldable for it.

Reply:
`Module` I think also gets different `objectid` in different sessions?

Reply:
`Module` is hashed by name and parent name as of https://github.com/JuliaLang/julia/pull/47749. Good catch on `F`. I think the right test is valid type param